Huffman Code question generator

Fabianstw opened the pull request:
I added a question generator for Huffman-Coding. 
There a currently 3 different variants for the user (only two are implemented)
1. choice option
- It generates words with a length of 8-13 chars. Those words have a unique Huffman-Tree generation. It also generates wrong answers, between which the user can choose
2.  input option
- like the choice option, but instead of wrong answers this is a free text question 
3. choice2 (not fully implemented) 
- here it currently creates longer options for Huffman coding, but instead of showing the user the word, it shows the char with the corresponding frequency (as dict)
- Still needs a better interface 
- Missing the generation of wrong answers

AbcAeffchen commented:
Thanks for the PR. I only had time for a quick look into the code. A few issues I noticed:
- the `.idea` folder shouldn't be in the repo. It is also in the  .gitignore file so I'm not sure what went wrong here.
- the generator is named `HuffmanCodingMultipleChoise` but it is used for generating any type of question. Following the naming of the other generators it should be only `HuffmanCoding` or something like that.
- all the huffman related files names are prefixed with "huffmanCoding" which is redundant since they are also in the huffman-coding folder and thus can be shortend.
- For the varaints: If I understand it correctly, `choice` and `choice2` only differ in the way the question is formulated, so if it is presented as a word or a table containing letter frequencies. Instead of heaving it called `choice2` this should be a boolean parameter switching between these two ways of presenting the input.

I comment on some other things in the code and need some time to actually have a look at the generated questions.

Fabianstw commented:
I removed the `.idea` folder, renamed the QuestionGenerator from `HuffmanCodingMultipleChoise` to `HuffmanCoding`, removed `huffmanCoding` from most files in the corresponding folder and I added a probability to differ between choice and choice2. So only two variants are visible to the user.

holgerdell commented on shared/src/question-generators/huffman-coding/huffmanCodingGenerateWrongAnswers.ts:
```diff
@@ -0,0 +1,203 @@
+  let current_longest_key: string = ""
+  for (const current_key in huffmanDict) {
+    if (current_longest_key === "") {
+      current_longest_key = current_key
```
Should we enable this rule? https://eslint.org/docs/latest/rules/camelcase

AbcAeffchen replied:
Probably yes. I don't see a down side of this. Maybe with the `ignoreImports: true` option so we don't need to rename imports of external modules just to fit the naming.

From the examples, it looks like it still allows for underscore prefix, and SCREAMING_SNAKE_CASE variables, which is fine. It also seems to allow snake_case function names which I would avoid.

AbcAeffchen replied:
ok, I will add it later. And when I'm on it I think I want to change prettier to a `printWidth` of 110. This worked fine for me the last weeks and prevents a lot of short lines split into multiple lines.

Fabianstw commented on shared/src/question-generators/huffman-coding/huffmanCodingGenerateWrongAnswers.ts:
```diff
@@ -0,0 +1,203 @@
+    "",
+  )
+  const keySet: string[] = Object.keys(huffmanDict)
+  const [randomKey1, randomKey2] = keySet
```
I changed the code to use the `random.subset` function. It is much more simple your way 👍

Fabianstw commented on shared/src/question-generators/huffman-coding/huffmanCodingAlgorithm.ts:
```diff
@@ -0,0 +1,179 @@
+  /**
+   * Function to get the frequency of all leafs
+   */
+  public getLeafs(leafs: number[]) {
```
I implemented this function for generating wrong answers, but in the end I didn't knew were I should use it. It seems redundant for me now, so I deleted it

holgerdell commented on shared/src/question-generators/huffman-coding/huffmanCoding.ts:
```diff
@@ -218,29 +218,25 @@ export const HuffmanCoding: QuestionGenerator = {
-      const choice1orchoice2 = random.uniform();
-      if (choice1orchoice2 >= 0.5) variant = 'choice2'
+      const choice1OrChoice2 = random.uniform();
+      if (choice1OrChoice2 >= 0.5) variant = 'choice2'
```
Es ist evtl. einfacher, stattdessen das hier zu benutzen:
```
variant = random.choice(["choice", "choice2"])
```

AbcAeffchen commented on shared/src/question-generators/example/example.ts:
```diff
@@ -28,7 +28,7 @@ function generateWrongAnswers(
   correctAnswer: number,
 ): Array<string> {
-  let wrongAnswers = [
+  const wrongAnswers = [
```
Just as a side note: this change is correct but does not belong into this branch/commit. Just leave it in for now but in general try to keep the branches/commits focused.

Fabianstw replied:
Okay, I will write it down in the future and then do those commits separately 👍

holgerdell commented on shared/src/question-generators/huffman-coding/huffManCodingGenerateWords.ts:
```diff
@@ -11,120 +11,125 @@ import Random from "@shared/utils/random.ts";
+    return ""
+  }
+
+  const tmpFrequencies: { [key: number]: number[][] } = {
```
If this dictionary never changes, it could be defined outside of the function scope.

holgerdell replied:
The type of `length` could then be `keyof tmpFrequencies`

holgerdell commented on shared/src/question-generators/huffman-coding/huffManCodingGenerateWords.ts:
```diff
@@ -11,120 +11,125 @@ import Random from "@shared/utils/random.ts";
+
+  // create the word based on the chosen frequencies
+  for (let i = 0; i < chosenFrequency.length; i++) {
+    const char = possibleChars[random.int(0, possibleChars.length - 1)]
```
das ist gefährlich, weil der selbe Buchstabe mehrmals gewählt werden könnte. Du meinst glaube ich
`random.subset(possibleChars, chosenFrequency.length)` oder ähnlich.

holgerdell commented on shared/src/question-generators/huffman-coding/huffManCodingGenerateWords.ts:
```diff
@@ -11,120 +11,125 @@ import Random from "@shared/utils/random.ts";
+  }
+  const charArray: { [key: string]: number } = {}
+  amountCharsArray.forEach((item) => {
+    const randomLetter = random.choice(possibleChars.split(""))
```
auch hier wird der selbe Buchstabe ggf. mehrfach gewählt. `random.subset`?

holgerdell commented:
@Fabianstw Kannst du bitte
- vorher sichern `git checkout -b just-in-case`
- einmal `git rebase upstream/main --squash` durchführen, damit die Änderung atomar ist und auch leichter kommentiert werden kann
- dann `git push --force-with-lease` in dein Repo, dadurch wird der pull request automatisch auf den neuen commit aktualisiert.

AbcAeffchen commented:
> @Fabianstw Kannst du bitte
> 
>     * vorher sichern `git checkout -b just-in-case`
> 
>     * einmal `git rebase upstream/main --squash` durchführen, damit die Änderung atomar ist und auch leichter kommentiert werden kann
> 
>     * dann `git push --force-with-lease` in dein Repo, dadurch wird der pull request automatisch auf den neuen commit aktualisiert.

You mean for the review of the PR? Otherwise merging into a branch in this repo has the option to squash.

holgerdell commented on shared/src/question-generators/huffman-coding/huffmanCodingGenerateWrongAnswers.ts:
```diff
@@ -0,0 +1,203 @@
+  correctAnswer: string,
+) {
+  const wrongAnswer = correctAnswer.split("")
+  while (wrongAnswer.join("") === correctAnswer) {
```
wie wär's stattdessen mit `const flipPositions = random.subset(Array.from({length: n}, (_, i) => i), numFlips)`, um die Positionen für die Flips zu wählen. Dann braucht man die while-Schleife nicht.

Fabianstw replied:
Habe ich ähnlich übernommen, aber für `numFlips` habe ich `random.int(1, 5)` genommen, damit es zufälliger ist.

holgerdell commented:
> You mean for the review of the PR? Otherwise merging into a branch in this repo has the option to squash.

Ja, ich meinte zum besseren PR review - die Zwischenversionen stören dafür. Evtl. ist das auch eher eine Bitte für die Zukunft.

AbcAeffchen commented:
> > You mean for the review of the PR? Otherwise merging into a branch in this repo has the option to squash.
> 
> Ja, ich meinte zum besseren PR review - die Zwischenversionen stören dafür. Evtl. ist das auch eher eine Bitte für die Zukunft.

Ich bin auch noch nicht sicher wie der Workflow gedacht ist. Mit `force-push` habe ich schlechte Erfahrungen gemacht, auch wenn es in dem speziellen Fall vielleicht geht. evtl. sind dann aber auch alle Kommentare weg?

Die Kommentare zum Review erscheinen ja hier, d.h. dafür muss man nicht nochmal in den Code schauen. Und man kann sie als Resolved markieren (wenn es dann soweit ist) dann werden sie ausgeblendet. Und wenn man oben auf "files changed" click" sieht man die summe aller commits, aber ggf. nicht die Kommentare die schon existieren, deren Codezeile aber nicht mehr existiert.

Vielleicht hilft etwas davon das Review einfacher zu machen?

holgerdell commented:
Danke, `files changed` ist, wonach ich gesucht hatte. Passt so wie es ist.

Fabianstw commented on shared/src/question-generators/huffman-coding/huffmanCodingAlgorithm.ts:
```diff
@@ -0,0 +1,179 @@
+/**
+ * This class represents a node in the huffman tree
+ */
+export class TreeNode {
```
Habe `TreeNode` geändert zu `export type TreeNode = {...}`, dadurch ist der Code auch übersichtlicher geworden.

Fabianstw commented on shared/src/question-generators/huffman-coding/huffmanCodingAlgorithm.ts:
```diff
@@ -0,0 +1,179 @@
+  random: Random,
+) {
+  const firstValue = random.int(0, 1)
+  const secondValue = (firstValue + 1) % 2
```
Das stimmt! Habe ich geändert :-)

Fabianstw commented on shared/src/question-generators/huffman-coding/huffManCodingGenerateWords.ts:
```diff
@@ -11,120 +11,125 @@ import Random from "@shared/utils/random.ts";
+  random: Random,
+): string {
+  if (difficulty === 0) {
+    return ""
```
Ich habe vorerst `difficulty` rausgenommen, da ich noch keine Anwendung dafür gefunden habe. Bei Bedarf können wir es relativ schnell wieder implementieren.

Fabianstw commented on shared/src/question-generators/huffman-coding/huffmanCoding.ts:
```diff
@@ -218,29 +218,25 @@ export const HuffmanCoding: QuestionGenerator = {
+    ];
+    const wordlength = random.weightedChoice(wordLengths);
 
     let variant : string = parameters.variant as "choice" | "input"
```
Hier habe ich den oberen Vorschlag 

> `variant = random.choice(["choice", "choice2"])` 
übernommen und den Typ `:string` entfernt. 

Ich bin mir aber nicht ganz sicher ob es jetzt korrekt/besser ist.

holgerdell replied:
Ich auch nicht. Ich glaube, `as "choice" | "input"` sollte wohl auch weg gelassen werden, da unnötig.

`random.choice(..)` ist besser als das was vorher da stand.

holgerdell commented on front-end/src/components/DrawTable.tsx:
```diff
@@ -0,0 +1,88 @@
+import { ReactElement } from "react" // , useEffect, useRef
```
please remove the comment

holgerdell commented on front-end/src/components/DrawTable.tsx:
```diff
@@ -0,0 +1,88 @@
+      <tr>
+        {parsedHeader[i].map((x) => (
+          // eslint-disable-next-line react/jsx-key
+          <th style={cellStyle}>{x}</th>
```
wir benutzen tailwind anstatt explizitem CSS. Also durch className="p-2 border-1" o.ä. ersetzen. Die Border-Farbe ist bereits automatisch vom Theme als CSS-Variable gesetzt worden und muss hier nicht explizit angegeben werden.

holgerdell commented on front-end/src/components/DrawTable.tsx:
```diff
@@ -0,0 +1,88 @@
+      <tr>
+        {parsedContent[i].map((x) => (
+          // eslint-disable-next-line react/jsx-key
+          <td style={cellStyle}>{x}</td>
```
Eine Eigenheit von React ist, dass im Zusammenhang mit map am element ein key angegeben werden muss, damit die Daten dem Element zugeordnet werden können. Das ist wichtig, sonst passieren manchmal komische Sachen.

Da x nicht eindeutig ist würde ich als key einfach den Spaltenindex nehmen, ich glaube das müsste gehen:
```
{parsedContent[i].map((x, j) => (<td key={`${j}`} style={cellStyle}>{x}</td>)
```

holgerdell commented on front-end/src/components/DrawTable.tsx:
```diff
@@ -0,0 +1,88 @@
+
+  return (
+    <table>
+      <thead>{tableHeader.map((x) => x)}</thead>
```
das müsste auch ohne map gehen? Also einfach {tableHeader}.

holgerdell commented on front-end/src/components/DrawTable.tsx:
```diff
@@ -0,0 +1,88 @@
+  return {
+    header: header,
+    content: content,
+  }
```
hier kann man auch kürzer `return {header, content}` schreiben.

holgerdell commented on front-end/src/components/Markdown.tsx:
```diff
@@ -87,6 +88,13 @@ export const MarkdownTreeNode: FunctionComponent<{
+          <DrawTable table={parseTreeNode.child} />
+        </div>
+      )
+    }
```
Das ist nicht falsch, aber sauberer wäre es, wenn wir hier der normalen Markdown-Syntax für Tabellen folgen:
https://www.markdownguide.org/extended-syntax/#tables
Also ohne, dass man in einen Code-Block rein muss. Dazu muss `parseMarkdown.ts` angepasst werden, die `parseTable`-Funktion sollte auch dort rein.

holgerdell commented on front-end/src/components/DrawTable.tsx:
```diff
@@ -0,0 +1,88 @@
+
+/**
+ * A component that returns an SVG element representing a graph.
+ * @param width - The width of the svg.
```
Kommentar anpassen

holgerdell commented on front-end/src/components/DrawTable.tsx:
```diff
@@ -0,0 +1,88 @@
+    const isHeader = contentSplit.every((x) => x === "-")
+    if (isHeader) {
+      if (matchedHeader) {
+        throw new Error("Table header can only be defined once")
```
ist ziemlich streng, ich würde es einfach so interpretieren, dass alle dem Header-Trenner folgenden Zeilen eben diese -- Einträge haben. Auch sollte die |--|--| Zeile nur dann als Header-Trenner interpretiert werden, wenn sie die zweite Zeile ist.

holgerdell commented on front-end/src/components/DrawTable.tsx:
```diff
@@ -0,0 +1,88 @@
+
+  for (let i = 0; i < rows.length; i++) {
+    // check if the row is like |-|-|-| or |--|--|--|
+    const contentSplit = rows[i].split("|").filter((x) => x !== "")
```
Würde jetzt

|||||||

nicht als Trenner interpretiert? Laut https://www.markdownguide.org/extended-syntax/#tables muss jede Zelle mindestens drei - enthalten. Und `|--- ---|---|` ist wahrscheinlich auch verboten.

Fabianstw commented:
> @Fabianstw Du kannst reviews auch als resolved markieren sollte sich der jeweilige Kommentar durch einen neuen commit erledigt haben. Nur falls du das nicht eh schon machst.

Ja habe ich teilweise schon gemacht, aber werde nachher nochmal genauer durchgehen 👍

holgerdell commented on shared/src/question-generators/huffman-coding/huffmanCoding.ts:
```diff
@@ -0,0 +1,446 @@
+    description: "Compute the Huffman-Coding of a given string",
+    text: 'Let "*{{0}}*" be {{1}}. What is a correct **Huffman-Coding** of this {{1}}?',
+    prompt: "What is a possible Huffman-Coding?",
+    textTable: `Suppose we have the following table, which represents how often a char appears in a string:
```
replace with `how often each character appears in a given string`

holgerdell commented on shared/src/question-generators/huffman-coding/huffmanCoding.ts:
```diff
@@ -0,0 +1,446 @@
+    description: "Bestimme die Huffman-Codierung eines gegebenen Strings",
+    text: 'Sei "*{{0}}*" ein {{1}}. Was ist eine korrekte **Huffman-Codierung** für diesen {{1}}?',
+    prompt: "Was ist eine mögliche Huffman-Codierung?",
+    textTable: `Angenommen wir habe die folgende Tabelle, welche angebibt, wie oft ein bestimmter Buchstabe in einem String vorkommt:
```
`wie oft jeder Buchstabe in einem gegebenen String vorkommt:`

holgerdell reviewed (approved):
Ich hab nur zwei Mini-Anmerkungen, ansonsten denke ich, dass wir das so mergen können (gerne mit rebase+squash im Github pull-request Interface).

holgerdell commented:
> > @Fabianstw Kannst du bitte
> > ```
> > * vorher sichern `git checkout -b just-in-case`
> > 
> > * einmal `git rebase upstream/main --squash` durchführen, damit die Änderung atomar ist und auch leichter kommentiert werden kann
> > 
> > * dann `git push --force-with-lease` in dein Repo, dadurch wird der pull request automatisch auf den neuen commit aktualisiert.
> > ```
> 
> You mean for the review of the PR? Otherwise merging into a branch in this repo has the option to squash.

Mach was auch immer sinnvoll erscheint und einen merge commit nach Möglichkeit vermeidet. Ich habe keine starke Meinung dazu.
